lib/Sema/CSDiag.cpp

```diff
@@ -4264,8 +4264,14 @@ bool TypeChecker::diagnoseSelfAssignment(const Expr *E) {
   if (!AE)
     return false;
 
+  // Allow the use of parentheses to express self-assignment when it's a deliberate choice
+  auto src = AE->getSrc();
+  auto loadSrc = dyn_cast<LoadExpr>(src);
```

Comment:
Yes, you wouldn't *expect* to be doing an assignment to a `let`, but someone may try to and I was operating under the assumption that we would reach this point if they did. I just checked and it turns out we won't hit this point for reasons that have to do with type checker implementation details.

It appears that the reason we started diagnosing expressions with parens on the RHS is because in `findReferencedDecl()` we are now using `getValueProvidingExpr()`. One simple fix might be to bail specifically on `ParenExpr` before that point. That makes this less dependent on the `ParenExpr` being nested under the `LoadExpr`, but also means that we'll stop digging into expressions as soon as we see that `ParenExpr`.

Related to that, I notice from the JIRA is that we are diagnosing `(i = i)`. That is because we are explicitly testing for `AssignExpr` in `diagnoseSelfAssignment` and not, e.g., using `getValueProvidingExpr()` and checking to see if that's an assignment.

I also notice that we will not diagnose something silly like this, despite the fact that there appears to be code for handling nested assignments in `findReferencedDecl()`:
```
func f() {
  var x: ()
  var i: Int = 5
  x = i = i
}
```

I'm not sure how much of this you want to look into and/or try to perfect. Your change is probably okay as is, but it might be slightly better to check for `ParenExpr` in `findReferencedDecl()` and bail out there for the time being. The reason this seems like a better choice to me is that I have always found it odd that we nest the `ParenExpr` under the `LoadExpr`.